doc/developer-guide.md

```diff
@@ -131,11 +137,59 @@ A CheckResult has the following properties:
 
 ### Common Functions
 
-Common functions are an internal library used by the rules and checks.  If you have code repeated across rules and checks, you can use these functions and contribute to them.  They are made available to every function as `commons`.  Documentation is available in [source code](../lib/commons/).
+Common functions are an internal library used by the rules and checks.  If you have code repeated across rules and checks, you can use these functions and contribute to them.  They are made available to every function as `commons`. Documentation is available in [source code](../lib/commons/).
+
+#### Commons and Shadow DOM
+
+To support Shadow DOM while maintaining backwards compatibility, commons functions that
+query DOM nodes must operate on an in-memory representation of the DOM using aXe-core’s
+built-in [API methods and utility functions](./API.md#virtual-dom-utilities).
+
+Commons functions should do the virtual tree lookup and call a `virtual` function
+including the rest of the commons code. The naming of this special function
+should contain the original commons function name with `Virtual` added to signify
+it expects to operate on a virtual DOM tree.
+
+Let’s look at an example:
+
+```javascript
+axe.commons.text.accessibleText = function (element, inLabelledbyContext) {
+  let virtualNode = axe.utils.getNodeFromTree(axe._tree[0]); // throws an exception on purpose if axe._tree not correct
+  return axe.commons.text.virtualAccessibleText(virtualNode, inLabelledbyContext);
+}
+
+axe.commons.text.virtualAccessibleText = function (element, inLabelledbyContext) {
+  // rest of the commons code minus the virtual tree lookup, since it’s passed in
+}
+```
+
+`virtualAccessibleText` would never be called directly; rather, it would be called
```

Comment:
That's not true. We always call the virtual functions if we've got a virtual node we can use. Only if we don't have one do we use the lookup function.